### PR TITLE
Root-only query handling

### PR DESCRIPTION
**Pull request by V0ldek**
## Short description

The primary goal is to support the empty query on arbitrary JSON documents to behave as one would expect.

```console
$ rq '$' ./empty_file.json -r count
0
```

```console
$ rq '$'
$      {}          
{}
```

```console
$ rq '$'
$ true
true
```

To facilitate that, we move the empty query special case over to a module and specialize all result modes. Most of them are really simple &ndash; asking for count is just asking if the root exists (count of 1) or not (count of 0) and can be done without looking at most of the input. All of those specializations assume valid input JSONs.

As an apropos fix, internal block padding was changed from null bytes to spaces &ndash; null bytes should not appear in a valid JSON at any point, and whitespace are properly neutral to the engine.

## Issue

Resolves: #160

## Checklist

All of these should be ticked off before you submit the PR.

- [x] I ran `just verify` locally and it succeeded.
- [x] Issue was given <span style="color: #FF4400">go ahead</span> and is linked above **OR** I have included justification for a minor change.
- [x] Unit tests for my changes are included **OR** no functionality was changed.

### CODE REVIEW

**Review by charles-paperman — APPROVED** [review]
Look through it quickly.  Sounds good.

File: crates/rsonpath-lib/src/input.rs [hunk 1]
```diff
@@ -169,7 +180,7 @@ pub(super) mod in_slice {
 
     #[inline]
     pub(super) fn pad_last_block(bytes: &[u8]) -> LastBlock {
-        let mut last_block_buf = [0; MAX_BLOCK_SIZE];
+        let mut last_block_buf = [b' '; MAX_BLOCK_SIZE];
```
Comment (zwerdlds) [inline crates/rsonpath-lib/src/input.rs:183]: I'd consider moving this to a constant, if possible.

Comment (zwerdlds) [reply crates/rsonpath-lib/src/input.rs:183]: Though I guess for a test it's ok.

File: crates/rsonpath-lib/src/lib.rs [hunk 2]
```diff
@@ -346,3 +346,9 @@ impl<F: FallibleIterator> Iterator for FallibleIntoIter<F> {
         }
     }
 }
+
+#[inline(always)]
+#[must_use]
+pub(crate) fn is_json_whitespace(x: u8) -> bool {
+    x == b' ' || x == b'\t' || x == b'\n' || x == b'\r'
```
Comment (zwerdlds) [inline crates/rsonpath-lib/src/lib.rs:353]: No opinion on perf, but legibility here *might* be improved with a match

**Review by zwerdlds — COMMENTED** [review]
Just a couple of style/legibility things for thinking on

**Pull request merged**